Fix the assertion failure occurred

thoshiai opened the pull request:
When postgres build with --enable-cassert, creating matview with
IVM or updating base table was failed.

I checked regression test is success with using --enable-cassert.
The assertion failure suppressed by this fix. But same case possibly exist.

thoshiai committed:
Fix the assertion failure occurred

When postgres build with --enable-cassert, to create matview with
IVM or to update base table was failed.

yugo-n commented on src/backend/commands/createas.c:
```diff
@@ -348,6 +348,8 @@ ExecCreateTableAs(CreateTableAsStmt *stmt, const char *queryString,
 			FuncCall *fn;
 
+			pstate->p_expr_kind = EXPR_KIND_SELECT_TARGET;
+
```
It seems good to me. 

We might be able to make a function to initialize pstate for our purpose, but I think this fix is enough for now.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -1055,7 +1058,24 @@ IVM_immediate_maintenance(PG_FUNCTION_ARGS)
+	 * The stored query was rewritten at the time of the MV definition, but
+	 * has not been scribbled on by the planner.
+	 */
+	query = linitial_node(Query, actions);
```
Great! I originally consider to fix the Assert condition in get_view_query(), but your solution that doesn't use get_view query and get it by itself seems better because  ExecRefreshMatView() does in this manner.

By the way, is there any reason or intention to check internal errors of only two condition?  ExecRefreshMatView() checks other conditions, for example ""materialized view \"%s\" is missing rewrite information".  If there is no problem, how about making a function to get a view definition query for matviews and calling this from both functions?

thoshiai replied:
I think that other check internal errors are unnecessary.
So no problem even if add. I make its function which get a view definition query for matview.

yugo-n replied:
I am just interested in why you think some checks are necessary and others are not. If there is good reason, I think it is nice to merge this fix. We can make the common function after we recognize making this is needed of better.

thoshiai replied:
sorry... I misunderstood. others checks are necessary too.

yugo-n reviewed (commented):
Thanks! 

There are some comments...

thoshiai committed:
Add get_materializedview_query() function.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -1749,3 +1666,79 @@ apply_delta(Oid matviewOid, Oid tempOid_new, Oid tempOid_old,
+ */
+Query *
+get_materializedview_query(Relation matviewRel, bool concurrent)
+{
```
This might be a minute thing, but how about get_materialized_view_query or get_matview_query?

thoshiai replied:
Ok, I will change function name as get_matview_query.

yugo-n commented on src/backend/commands/matview.c:
```diff
@@ -1749,3 +1666,79 @@ apply_delta(Oid matviewOid, Oid tempOid_new, Oid tempOid_old,
+	 * has not been scribbled on by the planner.
+	 */
+	return linitial_node(Query, actions);
+}
```
Hmm... I think this function should not include the check for unique index in CONCURRENT case because this part is not related to "getting view query". I guess the reason why this part precedes initial_node() is that the original author would like to do all checks which could raise errors earlier stage. Therefore, I think there is no problem even if the unique index check is performed after initial_node() call. What do you think of this?

thoshiai replied:
I think too.  It is correct as function stracture.
I will remove  CONCURRENT case in  get_matview_query function.

yugo-n commented on src/include/commands/matview.h:
```diff
@@ -34,4 +34,6 @@ extern bool MatViewIncrementalMaintenanceIsEnabled(void);
 
 extern Datum IVM_immediate_maintenance(PG_FUNCTION_ARGS);
 
+extern Query *get_materializedview_query(Relation matviewRel, bool concurrent);
```
This can be a static function instead of extern because this will be called only in matview.c.

thoshiai replied:
I will fix too.

thoshiai replied:
Are you not going to use this function at other place(for example createas.c) In the feature?

yugo-n replied:
I am not exactly sure , but I don't think so, especially as for createas.c, because when creating a materialized view ,the query information is available from the parsed statement. At least for now, this is not called from other place, so I think using static is better for good modularity. If we come to find the need to call this from outer in future, we can change this to use extern at that time.

yugo-n replied:
Although we might consider this when implementing ALTER MATERIALIZED VIEW if it is necessary and an appropriate manner....

thoshiai replied:
Ok,  If it will be need by other codes, we consider at that time.

thoshiai committed:
Fix get_materializedvieq_query()  to reflect comments.

Change function name from get_materializedview_query() to get_matview_query().
Change declaration to static function.
Reduce 'bool concurrent' argument.

thoshiai commented:
I fixed to reflect some comments.
nagata-san, Could you check this?

yugo-n commented:
OK. All tests passes. I'll merge this.
